File: src/lib/achievements.ts
import type { LucideIcon } from "lucide-react";
import { Film, Library, Trophy, Users } from "lucide-react";

export type AchievementCategory = "friends" | "completions" | "list";

export interface AchievementDefinition {
  id: string;
  category: AchievementCategory;
  threshold: number;
  title: string;
  description: string;
  icon: LucideIcon;
}

export interface AchievementUnlock {
  id: string;
  unlockedAt: string;
}

export interface ProfileAchievementStats {
  friendsCount: number;
  completionsCount: number;
  listCount: number;
}

export interface ProfileAchievementsState {
  stats: ProfileAchievementStats;
  unlocked: AchievementUnlock[];
}

export const ACHIEVEMENT_DEFINITIONS: AchievementDefinition[] = [
  {
    id: "friends_1",
    category: "friends",
    threshold: 1,
    title: "Primo amico",
    description: "Aggiungi il tuo primo amico.",
    icon: Users,
  },
  {
    id: "friends_5",
    category: "friends",
    threshold: 5,
    title: "Cerchia ristretta",
    description: "Hai 5 amici nella tua lista.",
    icon: Users,
  },
  {
    id: "friends_10",
    category: "friends",
    threshold: 10,
    title: "Popolare",
    description: "Hai 10 amici connessi.",
    icon: Users,
  },
  {
    id: "friends_25",
    category: "friends",
    threshold: 25,
    title: "Social star",
    description: "Hai 25 amici su Branchefy.",
    icon: Users,
  },
  {
    id: "watch_1",
    category: "completions",
    threshold: 1,
    title: "Prima visione",
    description: "Completa il tuo primo film, episodio o titolo.",
    icon: Film,
  },
  {
    id: "watch_5",
    category: "completions",
    threshold: 5,
    title: "Cinefilo",
    description: "Completa 5 titoli dall'inizio alla fine.",
    icon: Film,
  },
  {
    id: "watch_10",
    category: "completions",
    threshold: 10,
    title: "Maratoneta",
    description: "Completa 10 titoli.",
    icon: Film,
  },
  {
    id: "watch_25",
    category: "completions",
    threshold: 25,
    title: "Devoto seriale",
    description: "Completa 25 titoli.",
    icon: Film,
  },
  {
    id: "watch_50",
    category: "completions",
    threshold: 50,
    title: "Leggenda del divano",
    description: "Completa 50 titoli.",
    icon: Film,
  },
  {
    id: "list_1",
    category: "list",
    threshold: 1,
    title: "Da non perdere",
    description: "Salva il primo titolo nella tua lista.",
    icon: Library,
  },
  {
    id: "list_5",
    category: "list",
    threshold: 5,
    title: "Curatore",
    description: "Hai 5 titoli in lista.",
    icon: Library,
  },
  {
    id: "list_10",
    category: "list",
    threshold: 10,
    title: "Collezionista",
    description: "Hai 10 titoli in lista.",
    icon: Library,
  },
  {
    id: "list_25",
    category: "list",
    threshold: 25,
    title: "Archivista",
    description: "Hai 25 titoli in lista.",
    icon: Library,
  },
];

export const ACHIEVEMENT_BY_ID = Object.fromEntries(
  ACHIEVEMENT_DEFINITIONS.map((item) => [item.id, item]),
) as Record<string, AchievementDefinition>;

export const ACHIEVEMENT_CATEGORY_LABELS: Record<AchievementCategory, string> = {
  friends: "Amici",
  completions: "Visioni complete",
  list: "La mia lista",
};

export const ACHIEVEMENT_CATEGORY_ICONS: Record<AchievementCategory, LucideIcon> = {
  friends: Users,
  completions: Film,
  list: Library,
};

export function statForCategory(
  stats: ProfileAchievementStats,
  category: AchievementCategory,
): number {
  switch (category) {
    case "friends":
      return stats.friendsCount;
    case "completions":
      return stats.completionsCount;
    case "list":
      return stats.listCount;
    default:
      return 0;
  }
}

export function isWatchCompletedRatio(positionSecs: number, durationSecs?: number): boolean {
  if (durationSecs == null || durationSecs <= 0) return false;
  if (positionSecs <= 5) return false;
  return positionSecs / durationSecs >= 0.92;
}

export function streamingCompletionKey(input: {
  catalogPrefix: string;
  contentType: string;
  titleId: string;
  slug: string;
  videoId: string;
}): string {
  return `stream:${input.catalogPrefix.trim()}:${input.contentType.trim()}:${input.titleId.trim()}:${input.slug.trim()}:${input.videoId.trim()}`;
}

export function unlockedAchievementIds(unlocked: AchievementUnlock[]): Set<string> {
  return new Set(unlocked.map((item) => item.id));
}

export function nextAchievementInCategory(
  category: AchievementCategory,
  stats: ProfileAchievementStats,
  unlockedIds: Set<string>,
): AchievementDefinition | null {
  return (
    ACHIEVEMENT_DEFINITIONS.find(
      (item) =>
        item.category === category &&
        !unlockedIds.has(item.id) &&
        statForCategory(stats, category) < item.threshold,
    ) ?? null
  );
}

export function categoryProgress(
  category: AchievementCategory,
  stats: ProfileAchievementStats,
): { current: number; next?: AchievementDefinition } {
  const current = statForCategory(stats, category);
  const next = ACHIEVEMENT_DEFINITIONS.find(
    (item) => item.category === category && current < item.threshold,
  );
  return { current, next };
}

export function achievementProgressLabel(
  category: AchievementCategory,
  stats: ProfileAchievementStats,
): string {
  const { current, next } = categoryProgress(category, stats);
  if (!next) return `${current} completati`;
  return `${current} / ${next.threshold}`;
}

export const ACHIEVEMENTS_TAB_ICON = Trophy;

File: src-tauri/src/achievements.rs
use serde::Serialize;

use crate::db::Database;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AchievementUnlock {
    pub id: String,
    pub unlocked_at: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ProfileAchievementStats {
    pub friends_count: i32,
    pub completions_count: i32,
    pub list_count: i32,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ProfileAchievementsState {
    pub stats: ProfileAchievementStats,
    pub unlocked: Vec<AchievementUnlock>,
}

struct AchievementRule {
    id: &'static str,
    category: &'static str,
    threshold: i32,
}

const RULES: &[AchievementRule] = &[
    AchievementRule {
        id: "friends_1",
        category: "friends",
        threshold: 1,
    },
    AchievementRule {
        id: "friends_5",
        category: "friends",
        threshold: 5,
    },
    AchievementRule {
        id: "friends_10",
        category: "friends",
        threshold: 10,
    },
    AchievementRule {
        id: "friends_25",
        category: "friends",
        threshold: 25,
    },
    AchievementRule {
        id: "watch_1",
        category: "completions",
        threshold: 1,
    },
    AchievementRule {
        id: "watch_5",
        category: "completions",
        threshold: 5,
    },
    AchievementRule {
        id: "watch_10",
        category: "completions",
        threshold: 10,
    },
    AchievementRule {
        id: "watch_25",
        category: "completions",
        threshold: 25,
    },
    AchievementRule {
        id: "watch_50",
        category: "completions",
        threshold: 50,
    },
    AchievementRule {
        id: "list_1",
        category: "list",
        threshold: 1,
    },
    AchievementRule {
        id: "list_5",
        category: "list",
        threshold: 5,
    },
    AchievementRule {
        id: "list_10",
        category: "list",
        threshold: 10,
    },
    AchievementRule {
        id: "list_25",
        category: "list",
        threshold: 25,
    },
];

fn stat_for_category(stats: &ProfileAchievementStats, category: &str) -> i32 {
    match category {
        "friends" => stats.friends_count,
        "completions" => stats.completions_count,
        "list" => stats.list_count,
        _ => 0,
    }
}

pub fn get_state(
    db: &Database,
    profile_id: &str,
    cloud_friends_count: i32,
) -> Result<ProfileAchievementsState, String> {
    db.backfill_profile_completions(profile_id)?;
    let stats = achievement_stats(db, profile_id, cloud_friends_count)?;
    let unlocked = db
        .list_profile_achievements(profile_id)?
        .into_iter()
        .map(|(id, unlocked_at)| AchievementUnlock { id, unlocked_at })
        .collect();
    Ok(ProfileAchievementsState { stats, unlocked })
}

fn achievement_stats(
    db: &Database,
    profile_id: &str,
    cloud_friends_count: i32,
) -> Result<ProfileAchievementStats, String> {
    Ok(ProfileAchievementStats {
        friends_count: db.count_lan_friends(profile_id)? + cloud_friends_count.max(0),
        completions_count: db.count_profile_completions(profile_id)?,
        list_count: db.count_profile_list_items(profile_id)?,
    })
}

pub fn sync_profile(
    db: &Database,
    profile_id: &str,
    cloud_friends_count: i32,
) -> Result<Vec<AchievementUnlock>, String> {
    db.backfill_profile_completions(profile_id)?;
    let stats = achievement_stats(db, profile_id, cloud_friends_count)?;
    let mut newly_unlocked = Vec::new();

    for rule in RULES {
        let value = stat_for_category(&stats, rule.category);
        if value < rule.threshold {
            continue;
        }
        if let Some(unlocked_at) = db.unlock_profile_achievement(profile_id, rule.id)? {
            newly_unlocked.push(AchievementUnlock {
                id: rule.id.to_string(),
                unlocked_at,
            });
        }
    }

    Ok(newly_unlocked)
}

pub fn record_completion(
    db: &Database,
    profile_id: &str,
    completion_key: &str,
    kind: &str,
    title: &str,
    cloud_friends_count: i32,
) -> Result<Vec<AchievementUnlock>, String> {
    db.record_profile_completion(profile_id, completion_key, kind, title)?;
    sync_profile(db, profile_id, cloud_friends_count)
}
